### ISSUE TITLE
Passing training flag for batch normalization ?

### ISSUE THREAD

**Issue by unrealwill** [openai/spinningup#4]
Hello,

I am wondering how do you provide the training flag for batch normalization layers in your architecture specifying an actor critic function. 

During inference when generating the actions you want the training flag set to false, but when training you want it set to true.

This means that you need to use a different "Op" for generating the actions during the run of the episode, than when you train from the experience replay buffer.
But otherwise this "Op" for generating the actions must use exactly the same weights as the "main" network. 

So you either must carefully name your Ops, and use a variable scope with reuse=tf.AUTO_REUSE to make them share the weights, or recreate an Op and manage the copy from the "main" weights to the weights this new network before running a new episode.

How would you do it ?

**Comment by jachiam** [comment]
Hi,

Great question! To start off: batch norm is not supported out-of-the-box for actor-critic functions in Spinning Up. But if you wanted to hack it out, I think the natural approach would look like this: you would create a training flag placeholder in an algorithm's computation graph, and then pass that as an argument to the actor-critic function. Afterwards, you would modify feed_dicts, where appropriate, to also set the training flag.

The problem is that there may not be a super-principled approach to using it in a given algorithm. This is an under-explored topic in RL and I would encourage you to do research on it.

**Comment by jachiam** [comment]
For the Tensorflow `tf.layers.batch_normalization` layer, the `training` flag argument can be either a Python boolean **or** a TF boolean scalar tensor (eg placeholder). [source](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization)

Different RL algorithm implementations have different ways of addressing non-centered, non-unit-scaled inputs; I recommend checking out how Baselines does it if you're interested. Mostly it involves keeping running statistics for states and manually normalizing, in a way which always uses the running total statistics instead of batch statistics. But for simplicity's sake, the Spinning Up implementations omit this (and still do pretty well).

**Comment by unrealwill** [comment]
Sorry my mistake, you're right, you can indeed pass a placeholder.

I gave a look at https://github.com/openai/baselines/blob/master/baselines/common/mpi_running_mean_std.py , I find it more complex than using a layer which updates with a training flag, but I guess you can't do so because of mpi. Also btw (minor detail) but mathematically, keeping the running mean from the start probably spoil the convergence guarantees you usually have when using a properly initialized on-line update scheme (forgetting algorithm). 
To explain more clearly using mountain-car-v0 environment as example : during the first iterations the observations might move the mean, and their impact over time will decrease slowly in O(1/n) instead of decreasing exponentially. In mountain car the car position start at -0.5, when you start training you don't move a lot, so you will accumulate a lot of observation of ~ -0.5 position, which will then hinder the convergence of the mean of observation in the later stages of training.

**Comment by jachiam** [comment]
I don't think it's a minor detail: I actually personally find it very worrying that many deep RL implementations today use normalization schemes that break mathematical assumptions of the algorithms. It's not surprising that algorithms are robust to these kinds of hacks, but it breaks our intuition about what's working and why, which I think is harmful to research.

At any rate, I think this issue is resolved for now, so I'll mark it closed. Good luck in your experiments with normalization - let us know what you find in the end!
